node/src/components/contract_runtime/operations.rs

```diff
@@ -155,61 +133,111 @@ pub fn execute_finalized_block(
+    let execution_results = deploys
         .into_iter()
-        .exactly_one()
-        .map_err(|_| BlockExecutionError::MoreThanOneExecutionResult)?;
-    let json_execution_result = ExecutionResult::from(&ee_execution_result);
+        .filter_map(|deploy| {
+            let node_deploy_hash: NodeDeployHash = *deploy.id();
+            let deploy_hash = node_deploy_hash.into();
+            let deploy_header = deploy.take_header();
+            let ee_execution_result: EngineExecutionResult =
+                ee_exec_results.remove(&deploy_hash)?;
+            Some((
+                node_deploy_hash,
+                (deploy_header, (&ee_execution_result).into()),
+            ))
+        })
+        .collect::<BTreeMap<_, _>>();
+    Ok((execution_results, new_state_root_hash))
+}
 
-    let execution_effect: AdditiveMap<Key, Transform> = match ee_execution_result {
-        EngineExecutionResult::Success {
-            execution_journal,
-            cost,
-            ..
-        } => {
-            // We do want to see the deploy hash and cost in the logs.
-            // We don't need to see the effects in the logs.
-            debug!(?deploy_hash, %cost, "execution succeeded");
-            execution_journal
+fn extract_transforms_from_execution_results(
+    execution_results: &BTreeMap<DeployHash, EngineExecutionResult>,
+) -> AdditiveMap<Key, Transform> {
+    let mut transforms = AdditiveMap::new();
+    for (deploy_hash, ee_execution_result) in execution_results.iter() {
+        match ee_execution_result {
+            EngineExecutionResult::Success { cost, .. } => {
+                // We do want to see the deploy hash and cost in the logs.
+                // We don't need to see the effects in the logs.
+                debug!(?deploy_hash, %cost, "execution succeeded");
+            }
+            EngineExecutionResult::Failure { error, cost, .. } => {
+                // Failure to execute a contract is a user error, not a system error.
+                // We do want to see the deploy hash, error, and cost in the logs.
+                // We don't need to see the effects in the logs.
+                debug!(?deploy_hash, ?error, %cost, "execution failure");
+            }
         }
-        EngineExecutionResult::Failure {
-            error,
-            execution_journal,
-            cost,
-            ..
-        } => {
-            // Failure to execute a contract is a user error, not a system error.
-            // We do want to see the deploy hash, error, and cost in the logs.
-            // We don't need to see the effects in the logs.
-            debug!(?deploy_hash, ?error, %cost, "execution failure");
-            execution_journal
+        let journal = ee_execution_result.execution_journal().clone();
+        let journal: AdditiveMap<Key, Transform> = journal.into();
+        for (key, transform) in journal.iter() {
+            transforms.insert(*key, transform.clone());
```

Comment:
Isn't this going to overwrite the transform under the `key`? For `Write`s that's fine but not for `Add`s - we should merge the results.

Reply:
We need a test similar to [this one](https://github.com/dwerner/casper-node/blob/many-deploys-per-commit-3/execution_engine_testing/tests/src/test/regression/transforms_must_be_ordered.rs) but one that will include multiple deploys trying to modify the same keys (create transforms for the same key) and validate that the last comitted value is what we expect to be.

Reply:
Completed the ordering here, will look into the test you're suggesting.

Reply:
This is completed now.